pick moves using normal distribution

jiteshk23 opened the pull request:
Hey @singaurav,  just completed the first cut of pick moves using normal distribution. Pull the changes locally and try running on your machine. Let me know your thoughts.

jiteshk23 committed:
pick moves using normal distribution

singaurav commented:
@jiteshk23 I have an idea to make this generic. I think we can can generate a pdf(probability for each move) for move vector and write a generic function that takes a pdf and a count, and samples the moves. This way we will have complete runtime control over the sampling. And the RANDOM, NORMAL or any other sampling can be done by just generating the appropriate pdf. The pdf generating functions can also be made configurable. For ex we can pass mean and standard deviation to func which generates normal pdf. What say?

jiteshk23 commented:
Yeah sounds good. I'll try to implement this today.

jiteshk23 committed:
Add 2 functions - get_distribution and sample_indices

* currently not generic, normal distribution is hardcoded
* didn't find any base class for the various distributions in cpp
<random> to be able to abstract out the hardcoded normal distribution from
the aforementioned functions

jiteshk23 commented:
@singaurav I have created 2 functions out of the previous. But still unable to absract out the distribution. I am unable to find the base class for distributions in cpp <random> header. [This](https://stackoverflow.com/questions/32907254/uniform-random-distribution-base-class-for-both-int-and-double) might be of some help. I'll look into it. Will need to brush up on cpp templates I guess.

singaurav commented:
My suggestion would be to create a sub class of Move called SampleMove which has two extra fields, sample_prob(this is p of move being Sampled) and index(this is the index of the original move vector in GameInfo. Now we write vector<SampleMove> sample ( vector<SampleMove>, count) and then we can write functions like vector<SampleMove> uniform(vector<Move>), vector<SampleMove> normal(vector<Move>, mean, std dev). And the client can simply call one of the latter functions and pass the SampleMove vector to the first function to actually get the sample. Let me know what you think.

singaurav commented:
First function is
```vector<SampleMove> sample ( vector<SampleMove>, count) ``` and the latter functions are 
```vector<SampleMove> uniform(vector<Move>)``` and 
```vector<SampleMove> normal(vector<Move>, mean, std dev)```

jiteshk23 commented:
Actually I wanted to keep the functions get_distribution and sample_indices generic so that they can be used to sample any vector and not just `vector<Move>`. Would have eventually moved them to a Utilities class.

In terms of number of functions and client usage, it is similar to what you are suggesting
1. sample_indices(count, d)
2. d = get_normal_distribution(count)
3. d = get_uniform_distribution(count)

I think I have a solution now. I'll write a wrapper on top of 2 and 3.
get_distribution(name) // name could be 'UNIFORM', 'NORMAL' etc.

And then 'sample_moves' can specify a 'distribution_name' paramter.
```
sample_moves(......., distribution_name)
{
    .....
    distribution = get_distribution(distribution_name, count);
    .....
}
```

jiteshk23 committed:
created utils and refactored sample_moves functions

jiteshk23 commented on src/gameinfo.cpp:
```diff
@@ -162,94 +164,18 @@ unsigned GameWithSampledMoves::sample_count<PERC, double>(double perc) {
 
 template <>
 void GameWithSampledMoves::sample_moves<RANDOM>(unsigned count) {
```
@singaurav may be we should change RANDOM to UNIFORM

singaurav replied:
Agreed, uniform makes more sense.
